drake/solvers/NloptSolver.cpp

```diff
@@ -0,0 +1,287 @@
+#include "NloptSolver.h"
+#include "Optimization.h"
+
+namespace Drake {
+namespace {
+Eigen::VectorXd MakeEigenVector(const std::vector<double>& x) {
+  Eigen::VectorXd xvec(x.size());
+  for (size_t i = 0; i < x.size(); i++) {
+    xvec[i] = x[i];
+  }
+  return xvec;
+}
+
+TaylorVecXd MakeInputTaylorVec(const Eigen::VectorXd& xvec,
+                               const VariableList& variable_list) {
+  size_t var_count = 0;
+  for (const DecisionVariableView& v : variable_list) {
+    var_count += v.size();
+  }
+
+  auto tx = initializeAutoDiff(xvec);
+  TaylorVecXd this_x(var_count);
+  size_t index = 0;
+  for (const DecisionVariableView& v : variable_list) {
+      this_x.segment(index, v.size()) = tx.segment(v.index(), v.size());
+      index += v.size();
+    }
+  return this_x;
+}
+
+double EvaluateCosts(const std::vector<double>& x,
+                     std::vector<double>& grad,
+                     void* f_data) {
+  const OptimizationProblem* prog =
+      reinterpret_cast<const OptimizationProblem*>(f_data);
+
+  double cost = 0;
+  Eigen::VectorXd xvec = MakeEigenVector(x);
+
+  auto tx = initializeAutoDiff(xvec);
+  TaylorVecXd ty(1);
+  TaylorVecXd this_x(x.size());
+
+  if (!grad.empty()) {
+    grad.assign(grad.size(), 0);
+  }
+
+  for (auto const& binding : prog->getGenericObjectives()) {
+    size_t index = 0;
+    for (const DecisionVariableView& v : binding.getVariableList()) {
+      this_x.segment(index, v.size()) = tx.segment(v.index(), v.size());
+      index += v.size();
+    }
+
+    binding.getConstraint()->eval(this_x, ty);
+
+    cost += ty(0).value();
+    if (!grad.empty()) {
+      for (const DecisionVariableView& v : binding.getVariableList()) {
+        for (size_t j = v.index(); j < v.index() + v.size(); j++) {
+          grad[j] += ty(0).derivatives()(j);
+        }
+      }
+    }
+  }
+
+  return cost;
+}
+
+/// Structure to marshall data into the NLopt callback functions,
+/// which take only a single pointer argument.
+struct WrappedConstraint {
+  const Constraint* constraint;
+  const VariableList* variable_list;
+};
+
+double ApplyConstraintBounds(double result, double lb, double ub) {
+  // Our constraints are expressed in the form lb <= f(x) <= ub.
+  // NLopt always wants the value of a constraint expressed as
+  // f(x) <= 0.
+  //
+  // For upper bounds rewrite as: f(x) - ub <= 0
+  // For lower bounds rewrite as: -f(x) + lb <= 0
+
+  if (ub != std::numeric_limits<double>::infinity()) {
+    if ((lb != -std::numeric_limits<double>::infinity()) && (lb != ub)) {
+      throw std::runtime_error(
+          "Constraints with different upper and lower bounds not implemented.");
+    }
+    result -= ub;
+  } else {
+    if (lb == -std::numeric_limits<double>::infinity()) {
+      throw std::runtime_error(
+          "Unable to handle constraint with no bounds.");
+    }
+    result *= -1;
+    result += lb;
+  }
+  return result;
+}
+
+double EvaluateScalarConstraint(const std::vector<double>& x,
```

Comment:
Did you consider a comment explaining the math in this function and the next?  As-is I don't really follow (which might mean the next reader won't either).  I get this is all part of the NLOPT interface, so maybe a pointer to docs would suffice.